### PR TITLE
Spark

### PR DESCRIPTION
**Pull request by taragu**
Added Spark standalone, streaming examples, and integration with Cloud Pubsub

### CODE REVIEW

**Comment by taragu** [comment]
@lesv have you found any problems with my code? Thank you!

File: scala/spark-pubsub/README.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+section below), then transfer the application jar to a GCE VM. The GCE 
+VMs can be configured with bdutil to run Spark applications (link to 
+“Running the code sample with spark-submit”).
+
+
+## Before you start
+
+Before you run this code sample, you'll need to complete the following tasks:
+
+1. [Create a Cloud Bigtable cluster][create-cluster]. Be sure to note the
```
Comment (lesv) [inline scala/spark-pubsub/README.md]: Links are done by [text]\(link) There should be links to documentation or other explanation here.

File: scala/spark-pubsub/README.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+project ID.
+2. [Create a service account and a JSON key file][json-key].
```
Comment (lesv) [inline scala/spark-pubsub/README.md]: Also here.

**Comment by lesv** [comment]
@maxluebbe PTAL
@hegemonic PTAL

File: scala/spark-pubsub/README.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    //from https://github.com/taragu/cloud-bigtable-examples/blob/pubsub/scala/spark-pubsub-wordcount/cloud-pubsub-receiver/src/main/scala/CloudPubsubReceiver.scala#L97
+    val ackIDMessagesDStream = CloudPubsubUtils.createDirectStream(ssc, projectName, topicName, subscriptionName)
+
+CloudPubsubUtils.scala contains utility methods that can be used either in the Spark-Cloud Pubsub connector (listSubscriptions, and getClient) or the message processor (sendAcks, and createDirectStream). 
+
+RetryHttpInitializerWrapper.scala retries failed RPC calls, and is called in the getClient method in CloudPubsubUtils.scala
+
+You can use the Spark-Cloud Pubsub connector in two ways. If you would like to use it without any modification, you can download a pre-built jar from a GCS storage bucket with the following command:
+
+    $ gsutil cp gs://cloud-bigtable-examples/spark-cloud-pubsub-connector_2.10-0.0.jar PATH/TO/SAVE/THE/FILE
```
Comment (lesv) [inline scala/spark-pubsub/README.md]: @sduskis Is this really where / how we wish to distribute this?  Shouldn't we add it to our mvn repo?

Comment (lesv) [inline scala/spark-pubsub/README.md]: @sduskis @taragu @hegemonic After having a conversation in DevRel - we are recommending that we do not put jars in buckets, but in the GitHub Releases.

File: scala/spark-pubsub/README.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
+# Example: Integrating Spark Streaming with Cloud Pubsub
+
```
Comment (lesv) [inline scala/spark-pubsub/README.md]: Please put a Table of contents

File: scala/spark-pubsub/README.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+We need to build three applications: the Spark-Cloud Pubsub connector, the message producer, and the message processor. To build the Spark-Cloud Pubsub connector:
+
+    $ cd spark-cloud-pubsub-connector
+    $ sbt package
+
+You should see the connector jar in target/scala-2.10/. Copy it
+into a GCS bucket, log in to the master VM, then download the it from the GCS
+bucket. You can also download the pre-built connector instead of building it on your machine with the following command:
+
+    $ gsutil cp gs://cloud-bigtable-examples/spark-cloud-pubsub-connector_2.10-0.0.jar PATH/TO/SAVE/THE/FILE
```
Comment (lesv) [inline scala/spark-pubsub/README.md]: From the releases folder.

File: scala/spark-pubsub/README.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+
+
+## Running the code sample with spark-submit
+
+We need to run both the message producer and the message processor at the same time: the message producer publishes messages, and the message processor transforms the data and writes them to Cloud Bigtable. We suggest running the producer on your local desktop (if you don't want to install sbt on your GCE VMs), and running the processor on your VMs. Run the following commands to run your application with spark-submit.
+
+First, create a new topic in the Cloud Pubsub web UI. Please note the topic name.
+
+Next, open two terminals (A and B). In terminal A, we run the message producer; in terminal B, we run the message processor.
```
Comment (lesv) [inline scala/spark-pubsub/README.md]: shell / terminal windows / shell tabs

File: scala/spark-pubsub/README.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+In terminal B, log in to the master as user hadoop:
+
+    $ gcloud --project=[PROJECT_ID] compute ssh --zone=[ZONE] hadoop@[PREFIX]-m
+
+(Still in terminal B) you can download the message processor jars you submit in the previous section
+“Building the code sample”, or download a pre-compiled jar with the following
+command:
+
+	$ gsutil cp gs://cloud-bigtable-examples/cloud-pubsub-receiver_2.10-0.0.jar .
```
Comment (lesv) [inline scala/spark-pubsub/README.md]: github/gcp/cloud-bigtable-examples/releases (figure out real URL)

File: scala/spark-pubsub/README.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+(Still in terminal B) We also need to Cloud Pubsub API jar as well as the Spark-Cloud Pubsub connector on Spark's classpath in order to call their API in runtime. Download the Cloud Pubsub API jar and the connector jar with the following commands:
+
+    $ wget http://central.maven.org/maven2/com/google/apis/google-api-services-pubsub/v1-rev2-1.20.0/google-api-services-pubsub-v1-rev2-1.20.0.jar
+    $ gsutil cp gs://cloud-bigtable-examples/spark-cloud-pubsub-connector_2.10-0.0.jar .
```
Comment (lesv) [inline scala/spark-pubsub/README.md]: ditto

File: scala/spark-pubsub/README.md (hunk 3 above)
Comment (taragu) [inline scala/spark-pubsub/README.md]: @lesv How do I put the jar in the releases?

File: scala/spark-pubsub/README.md (hunk 1 above)
Comment (taragu) [inline scala/spark-pubsub/README.md]: I was able to open the links on my computer by clicking "Create a Cloud Bigtable cluster" from here https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-pubsub/README.md. Do I still need to change it to [text](link)?

Comment (lesv) [inline scala/spark-pubsub/README.md]: No -- it's a part of Markdown I guess I missed. :(

**Comment by lesv** [comment]
The JAR shouldn't be submitted here.

File: scala/spark-pubsub/cloud-pubsub-producer/cloudpubsubproducer.sbt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+scalaVersion := "2.10.3"
+
+libraryDependencies ++= Seq("org.apache.spark" % "spark-core_2.10" % "1.3.0", 
+      "org.apache.spark" % "spark-streaming_2.10" % "1.3.0",
+      "org.apache.hadoop" % "hadoop-common" % "2.5.2", 
+      "org.apache.hbase" % "hbase-client" % "1.0.1", 
+      "org.apache.hbase" % "hbase-server" % "1.0.1", 
+      "org.apache.hbase" % "hbase-common" % "1.0.1",
+      "org.apache.hbase" % "hbase-annotations" % "1.0.1",
+      "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317",
```
Comment (lesv) [inline scala/spark-pubsub/cloud-pubsub-producer/cloudpubsubproducer.sbt]: Highly unlikely this is correct.

File: scala/spark-pubsub/cloud-pubsub-producer/src/main/scala/CloudPubsubProducer.scala [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+/** Read an input file line by line, and publush each line 
+  * as a message to a Cloud Pubsub topic
+  */
+object CloudPubsubProducer {
+  def main(args: Array[String]) {
+    if (args.length < 3) {
+      throw new Exception("Please enter project name, topic name,"
+        + " and input file name as arguments. The project and "
+        + "topic must exist. Example: sbt \"project cloud-pubsub-"
+        + "producer\" \"run sduskis-hello-shakespear cmd_line_1 "
```
Comment (lesv) [inline scala/spark-pubsub/cloud-pubsub-producer/src/main/scala/CloudPubsubProducer.scala]: No way this is right.  Never have our names on stuff.

File: scala/spark-pubsub/cloud-pubsub-producer/cloudpubsubproducer.sbt (hunk 9 above)
Comment (taragu) [inline scala/spark-pubsub/cloud-pubsub-producer/cloudpubsubproducer.sbt]: What do you mean by submitting the jar here?

File: scala/spark-pubsub/cloud-pubsub-receiver/cloudpubsubreceiver.sbt [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+scalaVersion := "2.10.3"
+
+libraryDependencies ++= Seq("org.apache.spark" % "spark-core_2.10" % "1.3.0", 
+      "org.apache.spark" % "spark-streaming_2.10" % "1.3.0",
+      "org.apache.hadoop" % "hadoop-common" % "2.5.2", 
+      "org.apache.hbase" % "hbase-client" % "1.0.1", 
+      "org.apache.hbase" % "hbase-server" % "1.0.1", 
+      "org.apache.hbase" % "hbase-common" % "1.0.1",
+      "org.apache.hbase" % "hbase-annotations" % "1.0.1",
+      "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317",
```
Comment (lesv) [inline scala/spark-pubsub/cloud-pubsub-receiver/cloudpubsubreceiver.sbt]: 7.1.3 alpn is more likely.

File: scala/spark-pubsub/spark-cloud-pubsub-connector/sparkcloudpubsubconnector.sbt [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+scalaVersion := "2.10.3"
+
+libraryDependencies ++= Seq("org.apache.spark" % "spark-core_2.10" % "1.3.0", 
+		    "org.apache.spark" % "spark-streaming_2.10" % "1.3.0",
+		    "org.apache.hadoop" % "hadoop-common" % "2.5.2", 
+		    "org.apache.hbase" % "hbase-client" % "1.0.1", 
+		    "org.apache.hbase" % "hbase-server" % "1.0.1", 
+		    "org.apache.hbase" % "hbase-common" % "1.0.1",
+		    "org.apache.hbase" % "hbase-annotations" % "1.0.1",
+		    "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317",
```
Comment (lesv) [inline scala/spark-pubsub/spark-cloud-pubsub-connector/sparkcloudpubsubconnector.sbt]: 7.1.3 is more likely

File: scala/spark-pubsub/cloud-pubsub-receiver/cloudpubsubreceiver.sbt (hunk 11 above)
Comment (taragu) [inline scala/spark-pubsub/cloud-pubsub-receiver/cloudpubsubreceiver.sbt]: Bdutil uses 7.0.0: https://github.com/GoogleCloudPlatform/bdutil/blob/f1b5b36bae4bd8b5abb09d94d3575980b307755d/extensions/bigtable/bigtable_env.sh#L40
Should I use 7.1.3 instead even though it won't be in the VMs?

File: scala/spark-standalone-wordcount/README.md [hunk 13]
```diff
@@ -0,0 +1,20 @@
+# Running Spark in Standalone Mode with Cloud Bigtable
+
+This example uses Spark to perform a simple Spark job that counts the number of times a word appears in
+ a text file. The Spark job uses Cloud Bigtable to store the results.
+
+Spark: https://spark.apache.org/
+
+Please note that we encourage users to develop programs on their local machine, not on the VMs. User
```
Comment (lesv) [inline scala/spark-standalone-wordcount/README.md]: We probably don't need this sentence.  The rest of the paragraph is fine.

File: scala/spark-standalone-wordcount/README.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+Note that the bdutil release 1.3.1 does not have the latest configuration to connect Cloud Bigtable
+with Spark. Please use bdutil's git repository master branch to create and configure GCE VMs. 
+
+     $ git clone https://github.com/GoogleCloudPlatform/bdutil.git
+	 $ cd bdutil
+	 $ ./bdutil -e hadoop2_env.sh -e extensions/spark/spark_env.sh -e extensions/bigtable/bigtable_env.sh -e [path/to/config/file.sh] -f deploy
+
+Please go to the official documentation for specific instructions:
+TODO include link
```
Comment (lesv) [inline scala/spark-standalone-wordcount/README.md]: We should include more.

File: scala/spark-standalone-wordcount/src/main/scala/WordCount.scala [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+    // same conf object among the driver and executors
+    val confBroadcast = sc.broadcast(
+      new SerializableWritable(hbaseConfig))
+    // set config object to null to prevent it to be serialized 
+    // when using spark-shell
+    hbaseConfig = null
+    val conn = ConnectionFactory.createConnection(
+      confBroadcast.value.value)
+
+    // create new table if it's not already existed
```
Comment (lesv) [inline scala/spark-standalone-wordcount/src/main/scala/WordCount.scala]: Create a new table if it doesn't already exist.

File: scala/spark-standalone-wordcount/src/main/scala/WordCount.scala [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+    val wordCounts = sc.textFile(file).
+      flatMap(_.split(" ")).
+      filter(_!="").map((_,1)).
+      reduceByKey((a,b) => a+b)
+    wordCounts.foreachPartition {
+      partition => {
+        val config = confBroadcast.value.value
+	val conn1 = ConnectionFactory.createConnection(config)
+        val tableName1 = TableName.valueOf(name)
```
Comment (lesv) [inline scala/spark-standalone-wordcount/src/main/scala/WordCount.scala]: indentation

File: scala/spark-standalone-wordcount/src/main/scala/WordCount.scala [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+        val mutator = conn1.getBufferedMutator(tableName1)
+	try {
```
Comment (lesv) [inline scala/spark-standalone-wordcount/src/main/scala/WordCount.scala]: indentation ???

File: scala/spark-standalone-wordcount/wordcount.sbt [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
+scalaVersion := "2.10.3"
+
+libraryDependencies ++= Seq("org.apache.spark" % "spark-core_2.10" % "1.4.0", 
+		    "org.apache.hadoop" % "hadoop-common" % "2.5.1", 
+		    "org.apache.hbase" % "hbase-client" % "1.0.1.1", 
+		    "org.apache.hbase" % "hbase-server" % "1.0.1.1", 
+		    "org.apache.hbase" % "hbase-common" % "1.0.1.1",
+		    "org.apache.hbase" % "hbase-annotations" % "1.0.1.1",
+		    "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317")
```
Comment (lesv) [inline scala/spark-standalone-wordcount/wordcount.sbt]: 7.1.3 ???

File: scala/spark-streaming-wordcount/README.md [hunk 19]
```diff
@@ -0,0 +1,21 @@
+# Running Spark in Streaming Mode with Cloud Bigtable
+
+This example uses Spark streaming to pull new files from a GCS directory every 30 seconds and perform 
+a simple Spark job that counts the number of times a word appears in each new file. The Spark job uses
+Cloud Bigtable to store the results.
+
+Spark Streaming: http://spark.apache.org/docs/latest/streaming-programming-guide.html
+
+Please note that we encourage users to develop programs on their local machine, not on the VMs. User
```
Comment (lesv) [inline scala/spark-streaming-wordcount/README.md]: Remove the "please note" sentence.  I think I gave that to you as advice on how to write, not to put it in the readme -- please remove from all readme's.

File: scala/spark-streaming-wordcount/streamingexample.sbt [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+scalaVersion := "2.10.3"
+
+libraryDependencies ++= Seq("org.apache.spark" % "spark-core_2.10" % "1.4.0", 
+		    "org.apache.spark" % "spark-streaming_2.10" % "1.4.0",
+		    "org.apache.hadoop" % "hadoop-common" % "2.5.2", 
+		    "org.apache.hbase" % "hbase-client" % "1.0.1", 
+		    "org.apache.hbase" % "hbase-server" % "1.0.1", 
+		    "org.apache.hbase" % "hbase-common" % "1.0.1",
+		    "org.apache.hbase" % "hbase-annotations" % "1.0.1",
+		    "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317")
```
Comment (lesv) [inline scala/spark-streaming-wordcount/streamingexample.sbt]: 7.1.3 ??

**Comment by lesv** [comment]
Let me know when all of these are done.

**Comment by taragu** [comment]
@lesv bdutil uses ALPN 7.0.0: https://github.com/GoogleCloudPlatform/bdutil/blob/f1b5b36bae4bd8b5abb09d94d3575980b307755d/extensions/bigtable/bigtable_env.sh#L40
Should we use 7.1.3 instead even though it won't be in the VMs?

File: scala/spark-pubsub/cloud-pubsub-receiver/cloudpubsubreceiver.sbt (hunk 11 above)
Comment (lesv) [inline scala/spark-pubsub/cloud-pubsub-receiver/cloudpubsubreceiver.sbt]: @sduskis @taragu bdutil is wrong then.  alpn-boot needs to match the version of java you are using -- http://www.eclipse.org/jetty/documentation/current/alpn-chapter.html#alpn-versions

File: scala/spark-pubsub/cloud-pubsub-producer/cloudpubsubproducer.sbt (hunk 9 above)
Comment (lesv) [inline scala/spark-pubsub/cloud-pubsub-producer/cloudpubsubproducer.sbt]: 7.1.3 is likely the correct version -- bdutil is incorrect if it's using this.

File: scala/spark-standalone-wordcount/README.md (hunk 14 above)
Comment (taragu) [inline scala/spark-standalone-wordcount/README.md]: What should we include other than the link to the official documentation?

File: scala/spark-standalone-wordcount/src/main/scala/WordCount.scala (hunk 16 above)
Comment (taragu) [inline scala/spark-standalone-wordcount/src/main/scala/WordCount.scala]: Hmm I don't know why it shows up like this in the diff. It looks fine from here: https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-standalone-wordcount/src/main/scala/WordCount.scala.

**Comment by taragu** [comment]
@lesv I've addressed your comments. The indentation looks messed up in the diff but looks fine here: https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-streaming-wordcount/src/main/scala/StreamingExample.scala
https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-standalone-wordcount/src/main/scala/WordCount.scala

Also, you mentioned I should include more in the word count example README (https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-standalone-wordcount/README.md) besides the link to the official doc. What else should I include?

**Comment by sduskis** [comment]
I find that indentation issues happen when there are tabs in the document.

File: scala/spark-standalone-wordcount/src/main/scala/WordCount.scala (hunk 16 above)
Comment (lesv) [inline scala/spark-standalone-wordcount/src/main/scala/WordCount.scala]: Look for special characters in the file.  If you are happy w/ it, then I'll be.

File: scala/spark-standalone-wordcount/src/main/scala/WordCount.scala [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+    val name = args(1) //output table name
+    val expectedCount = args(2).toInt
+    val sc = new SparkContext()
+
+    var hbaseConfig = HBaseConfiguration.create()
+    hbaseConfig.set(TableInputFormat.INPUT_TABLE, name)
+    // broadcast a serialized config object allows us to use the
+    // same conf object among the driver and executors
+    val confBroadcast = sc.broadcast(
+      new SerializableWritable(hbaseConfig))
```
Comment (lesv) [inline scala/spark-standalone-wordcount/src/main/scala/WordCount.scala]: Should this be indent by 4?  I don't recall seeing anything in the guide, but most of our guides double the spacing for a carry on line.

**Comment by lesv** [comment]
Tara -- Look at Solomon's advice and remove any tabs or hidden characters.  Consider using 4 space indents for carry on lines.

I need to check w/ @hegemonic as we may just do a simple doc for this and have the bulk of the text in the various readme's.

**Comment by taragu** [comment]
@lesv I've removed the tabs. As we discussed earlier, I didn't use 4 spaces because the style guide used 2 spaces: https://sites.google.com/a/google.com/scala/coding/scala-style-for-new-developers?pli=1. Is there anything else?

**Comment by lesv** [comment]
Did you just strike the alpn-boot?  If so, does it work?

**Comment by taragu** [comment]
Yes---I deleted alpn in the build files because it was only used in runtime. We don't need it to compile the code.

**Comment by lesv** [comment]
LGTM -- please make sure to add it to the README about running, but lets start a new PR for that.

**Pull request merged**